### PR TITLE
Floating-Point Classification APIs

### PR DESCRIPTION
**Pull request by tannergooding**
This implements https://github.com/dotnet/corefx/issues/16841

### TODO
* The reference assemblies in CoreFX will need to be updated.
* CoreFX has a few calls to `IsInfinite || IsNaN` that can be replaced with `!IsFinite`
* CoreFX will need tests written

### CODE REVIEW

**Comment by tannergooding** [comment]
FYI. @mellinoe

**Comment by tannergooding** [comment]
What is the policy on attributing a method with `AggressiveInlining`? These APIs are small enough that the call is far more expensive than the actual code executed (each one, other than `IsNormal` and `IsSubnormal`, essentially compiles down to 1 or 2 asm instructions).

File: src/mscorlib/src/System/Double.cs [hunk 1]
```diff
@@ -44,56 +44,79 @@ public struct Double : IComparable, IFormattable, IConvertible
...
         }
 
+        /// <summary>Determines whether the specified value is NaN.</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
-        public static bool IsPositiveInfinity(double d)
+        public unsafe static bool IsNaN(double d)
         {
-            //Jit will generate inlineable code with this
-            if (d == double.PositiveInfinity)
```
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: I made this match the float implementation. I also distinctly remember some issue where this pattern actually generated poorer code than `return (d == double.PositiveInfinity)` (can't seem to find it right now).

File: src/mscorlib/src/System/Double.cs [hunk 2]
```diff
@@ -44,56 +44,79 @@ public struct Double : IComparable, IFormattable, IConvertible
...
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            return (bits & 0x7FFFFFFFFFFFFFFF) < 0x7FF0000000000000;
+        }
+
+        /// <summary>Determines whether the specified value is infinite.</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
         public unsafe static bool IsInfinity(double d)
         {
-            return (*(long*)(&d) & 0x7FFFFFFFFFFFFFFF) == 0x7FF0000000000000;
```
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: `var bits = BitConverter.DoubleToInt64Bits(d);` is significantly more readable and gets inlined to the same code (marking with `AgressiveInlining` would probably be beneficial though).

File: src/mscorlib/src/System/Double.cs [hunk 3]
```diff
@@ -44,56 +44,79 @@ public struct Double : IComparable, IFormattable, IConvertible
 
         internal static double NegativeZero = BitConverter.Int64BitsToDouble(unchecked((long)0x8000000000000000));
```
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: Side Note: Why isn't this public?

File: src/mscorlib/src/System/Double.cs [hunk 4]
```diff
@@ -44,56 +44,79 @@ public struct Double : IComparable, IFormattable, IConvertible
...
-                return false;
-            }
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            return (bits & 0x7FFFFFFFFFFFFFFF) > 0x7FF0000000000000;
         }
 
+        /// <summary>Determines whether the specified value is negative.</summary>
+        [Pure]
+        [System.Runtime.Versioning.NonVersionable]
+        public unsafe static bool IsNegative(double d)
```
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: this was `internal` previously.

**Comment by mellinoe** [comment]
It looks like some of the source files have the wrong encoding (GitHub thinks they are "binary files").

**Comment by tannergooding** [comment]
Fixed the encoding issue, seems one of the files (singlenan.cs) is encoded as UCS-2 LE

File: src/mscorlib/src/System/Double.cs (hunk 3 above)
Comment (mellinoe) [inline src/mscorlib/src/System/Double.cs]: Not sure how useful it is -- you can just use `-0.0`.

**Comment by AndyAyersMS** [comment]
For aggressive inlining-- generally one should assess the cost/benefit tradeoffs (code speed/code size/jit time) of forcing the inlines on some actual code. Since these are largely new methods that may be problematic.

My guess is they're not going to be called in enough places for any of the costs to add up, and since there is likely some marginal speed benefit from inlines, adding the aggressive inline attribute is fine.

**Comment by tannergooding** [comment]
Added `AggressiveInlining` attributes to all the methods except for `IsNormal` and `IsSubnormal`.

All of the functions to which it was added should compile down to roughly 2 instructions (an `AND` and a `CMP` instruction). `IsNormal` and `IsSubnormal` are more complex (roughly 5-8 instructions) and include branching, so they probably aren't worth inlining.

I also marked `BitConverter.Int64BitsToDouble` and related methods with `AggressiveInlining`, they should compile down to nothing since we are preserving the bit representation (so, provided `value` is already in a register, nothing needs to be done). I can revert this change if anyone feels strongly about it.

**Comment by tannergooding** [comment]
Looks like tests are failing because the reference assemblies don't contain the APIs yet. @mellinoe, is this another case where I need to check-in the implementation here, update the ref APIs in CoreFX, and then add tests back in CoreCLR?

**Comment by mellinoe** [comment]
@tannergooding Most likely, yeah.

File: src/mscorlib/src/System/BitConverter.cs [hunk 5]
```diff
@@ -339,21 +339,25 @@ public static bool ToBoolean(byte[] value, int startIndex)
             return value[startIndex] != 0;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This should compile down to nothing or roughly 1 instruction, depending on the register holding value
```
Comment (jkotas) [inline src/mscorlib/src/System/BitConverter.cs]: I don't think that this comment is right. Have you looked at the generated code?
I would expect this to compile to more than one instruction - spill the value to stack, fetch the value from stack as integer.

File: src/mscorlib/src/System/Double.cs [hunk 6]
```diff
@@ -44,56 +44,87 @@ public struct Double : IComparable, IFormattable, IConvertible
...
-            //Jit will generate inlineable code with this
-            if (d == double.PositiveInfinity)
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
+            var bits = BitConverter.DoubleToInt64Bits(d);
```
Comment (jkotas) [inline src/mscorlib/src/System/Double.cs]: I would expect the existing implementation to be faster than what you are replacing it with, on 32-bit architectures in particular.

File: src/mscorlib/src/System/BitConverter.cs (hunk 5 above)
Comment (tannergooding) [inline src/mscorlib/src/System/BitConverter.cs]: unless inlined, in which case the value should already be in a register and requires no translation (which is what I am seeing). I'll post the disasm tomorrow

Comment (tannergooding) [inline src/mscorlib/src/System/BitConverter.cs]: If forced to not inline on `BitConverter.DoubleToInt64Bits` and on `double.IsNaN`, the codegen is:

```asm
sub         rsp,28h
vzeroupper
vmovsd      xmm0,qword ptr [7FFD436104F0h]     ; double.IsNaN(double.NaN)
call        00007FFD43610090                   ;
```

**double.IsNaN**
```asm
sub         rsp,28h
vzeroupper
call        00007FFD43610420                   ; bits = BitConverter.DoubleToInt64Bits(d)
mov         rdx,7FFFFFFFFFFFFFFFh              ; bits &= 0x7FFFFFFFFFFFFFFF
and         rax,rdx                            ;
mov         rdx,7FF0000000000000h              ; return bits > 0x7FF0000000000000
cmp         rax,rdx                            ;
setg        al                                 ;
movzx       eax,al                             ;
add         rsp,28h
ret
```

**BitConverter.DoubleToInt64Bits**
```asm
vmovsd      qword ptr [rsp+8],xmm0             ; return *((long*)&value)
mov         rax,qword ptr [rsp+8]              ;
ret
```

If allowing inlining on `BitConverter.DoubleToInt64Bits` but forced to not inline on `double.IsNaN`, the codegen is:

```asm
sub         rsp,28h
vzeroupper
vmovsd      xmm0,qword ptr [7FFD436104A8h]     ; Load double.NaN
call        00007FFD43610090                   ; Call double.IsNaN
```

**double.IsNaN**
```asm
push        rax                                ;
vzeroupper                                     ;
vmovsd      qword ptr [rsp],xmm0               ; bits = *((long*)&value)
mov         rax,7FFFFFFFFFFFFFFFh              ; bits &= 0x7FFFFFFFFFFFFFFF
and         rax,qword ptr [rsp]                ;
mov         rdx,7FF0000000000000h              ; return bits > 0x7FF0000000000000
cmp         rax,rdx                            ;
setg        al                                 ;
movzx       eax,al                             ;
add         rsp,8
ret  
```

If allowing inlining on both, the codegen is:
```asm
sub         rsp,28h
vzeroupper
vmovsd      xmm0,qword ptr [7FFD435F04D0h]     ; bits *((long*)&value)
vmovsd      qword ptr [rsp+20h],xmm0           ;
mov         rcx,7FFFFFFFFFFFFFFFh              ; bits &= 0x7FFFFFFFFFFFFFFF
and         rcx,qword ptr [rsp+20h]            ;
mov         rax,7FF0000000000000h              ; result = bits > 0x7FF0000000000000
cmp         rcx,rax                            ;
setg        cl                                 ;
movzx       ecx,cl                             ;
```

Comment (tannergooding) [inline src/mscorlib/src/System/BitConverter.cs]: It seems like the JIT isn't optimizing the last case quite correctly, as it could be generating:
```asm
sub         rsp,28h
vzeroupper
mov         rcx,7FFFFFFFFFFFFFFFh              ; bits = *((long*)&value) & 0x7FFFFFFFFFFFFFFF
and         rcx,qword ptr [7FFD435F04D0h]      ;
mov         rax,7FF0000000000000h              ; result = bits > 0x7FF0000000000000
cmp         rcx,rax                            ;
setg        cl                                 ;
movzx       ecx,cl
```

File: src/mscorlib/src/System/Double.cs (hunk 6 above)
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: The existing implementation (which is manually inlining the BitConverter call) and the new implementation (which is more readable) produce the exact same ASM.

```asm
sub         rsp,28h  
vzeroupper  
vmovsd      xmm0,qword ptr [7FFD435F04D0h]
vmovsd      qword ptr [rsp+20h],xmm0  
mov         rcx,7FFFFFFFFFFFFFFFh  
and         rcx,qword ptr [rsp+20h]  
mov         rax,7FF0000000000000h  
cmp         rcx,rax  
setg        cl  
movzx       ecx,cl  
```

Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: Sorry, missed that you said on 32-bit in particular. x86 is also producing identical code between the two.

File: src/mscorlib/src/System/Double.cs [hunk 7]
```diff
@@ -44,56 +44,87 @@ public struct Double : IComparable, IFormattable, IConvertible
...
         }
 
+        /// <summary>Determines whether the specified value is NaN.</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
-        public static bool IsPositiveInfinity(double d)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // This compiles down to roughly 2 instructions
+        public unsafe static bool IsNaN(double d)
         {
-            //Jit will generate inlineable code with this
```
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: The JIT seems to special case comparisons against PositiveInfinity and NegativeInfinity, will fix this to be `return (d == double.PositiveInfinity)`

File: src/mscorlib/src/System/Double.cs (hunk 6 above)
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: Codegen on x86 could be improved by only comparing the upper 32-bits `return (upperbits & 0x7FFFFFFF) > 0x7FF00000`.

Comment (mikedn) [inline src/mscorlib/src/System/Double.cs]: > Codegen on x86 could be improved by only comparing the upper 32-bits return (upperbits & 0x7FFFFFFF) > 0x7FF00000.

It appears that you're using an older build or something, the current build produces:
```asm
G_M55887_IG02:
       C4E17B104508 vmovsd   xmm0, qword ptr [ebp+08H]
       C4E17B110424 vmovsd   qword ptr [esp], xmm0
       8B0424       mov      eax, dword ptr [esp]
       8B542404     mov      edx, dword ptr [esp+04H]
       83E0FF       and      eax, -1
       81E2FFFFFF7F and      edx, 0x7FFFFFFF
       83F801       cmp      eax, 1
       81DA0000F07F sbb      edx, 0x7FF00000
       0F9DC0       setge    al
       0FB6C0       movzx    eax, al
```

Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: I was using the 2.0 Preview 2 build, so that's entirely possible.

**Comment by jkotas** [comment]
> Looks like tests are failing because the reference assemblies don't contain the APIs yet.

Right. https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/adding_new_public_apis.md describes the workflow for adding new CoreLib APIs.

> add tests back in CoreCLR?

We prefer the API tests to be live in CoreFX (the doc says "add tests to the CoreFX test suite")

**Comment by tannergooding** [comment]
@jkotas, for previous FCALL APIs (Math, MathF), I have added minimal tests to CoreCLR (so-as to catch core regressions early) and added the bulk of the tests to CoreFX. I was planning on doing the same here.

**Comment by jkotas** [comment]
I do not think you need to bother with adding duplicated tests. Large part of CoreLib does not have tests in CoreCLR repo so adding a few more won't move the needle.

I would actually love to delete all tests under https://github.com/dotnet/coreclr/tree/master/tests/src/CoreMangLib and replace them by running a binary drop of subset of corefx tests.

**Comment by tannergooding** [comment]
@jkotas: https://github.com/dotnet/coreclr/issues/12782

File: src/mscorlib/src/System/Double.cs [hunk 8]
```diff
@@ -44,56 +44,85 @@ public struct Double : IComparable, IFormattable, IConvertible
...
         }
 
+        /// <summary>Determines whether the specified value is negative.</summary>
         [Pure]
         [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe static bool IsNegative(double d)
+        {
+            var bits = unchecked((ulong)BitConverter.DoubleToInt64Bits(d));
+            return (bits & 0x8000000000000000) == 0x8000000000000000;
```
Comment (jamesqo) [inline src/mscorlib/src/System/Double.cs]: Would this generate better code if it was `!= 0` instead? (I know you didn't write this, but just pointing it out.)

File: src/mscorlib/src/System/Single.cs [hunk 9]
```diff
@@ -40,38 +40,84 @@ public struct Single : IComparable, IFormattable, IConvertible
...
+
+        /// <summary>Determines whether the specified value is negative.</summary>
+        [Pure]
+        [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe static bool IsNegative(float f)
         {
-            return *(int*)(&f) == 0x7F800000;
+            var bits = unchecked((uint)BitConverter.SingleToInt32Bits(f));
+            return (bits & 0x80000000) == 0x80000000;
```
Comment (jamesqo) [inline src/mscorlib/src/System/Single.cs]: Same comment here

File: src/mscorlib/src/System/Double.cs (hunk 8 above)
Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: Potentially. I will double-check codegen (I actually had it as != 0 in my proposal, but there was an existing implementation, so I didn't update it past making it use BitConverter).

Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]: JIT recognizes this pattern and produces the same code as `!= 0`.

Comment (tannergooding) [inline src/mscorlib/src/System/Double.cs]:
```asm
sub         rsp,28h  
vzeroupper  
vmovsd      xmm0,qword ptr [7FF95BB004B8h]
vmovsd      qword ptr [rsp+20h],xmm0  
mov         rcx,8000000000000000h  
test        qword ptr [rsp+20h],rcx  
setne       cl  
movzx       ecx,cl
```

Code could be better (such as if the JIT supported the `BT` instruction), but that is out of scope for this PR.

**Comment by tannergooding** [comment]
Any other feedback here (or anything I missed)? I believe everything has been addressed.

**Comment by tannergooding** [comment]
Resolved merge conflicts.

**Comment by tannergooding** [comment]
@jkotas, could you do a final review pass (or ping the appropriate person to review/approve/merge)?

**Comment by jkotas** [comment]
It looks fine to me. I will leave it to @mellinoe as area-System.Numerics owner to merge and baby-sit through the system.

**Comment by tannergooding** [comment]
Thanks! I'll pick up the CoreFX side of things after we update the dependencies to include this change.

**Comment by jkotas** [comment]
@tannergooding You may need to submit the impl to CoreRT first (otherwise, you would need to ifdef out the CoreFX changes for CoreRT).

Bonus points for doing this by moving Single.cs and Double.cs to the shared partition:
1. Submit PR to CoreCLR to move Single.cs and Double.cs to `src/mscorlib/shared`
2. Once the CoreLib mirror creates a matching PR in CoreRT, push additional commits to it as necessary to make it build and work (delete the non-shared version of the files, etc.). It is usually a good idea to have these changes prepared upfront (e.g. like it is being done for TimeZone right now: https://github.com/dotnet/corert/pull/4188).

**Comment by tannergooding** [comment]
@jkotas, thanks for the heads up. I'll work on getting a PR created.

File: src/mscorlib/src/System/Double.cs [hunk 10]
```diff
@@ -44,56 +44,85 @@ public struct Double : IComparable, IFormattable, IConvertible
...
         [Pure]
-        internal unsafe static bool IsNegative(double d)
+        [System.Runtime.Versioning.NonVersionable]
+        // This is probably not worth inlining, it has branches and should be rarely called
+        public unsafe static bool IsNormal(double d)
         {
-            return (*(UInt64*)(&d) & 0x8000000000000000) == 0x8000000000000000;
+            var bits = BitConverter.DoubleToInt64Bits(d);
+            bits &= 0x7FFFFFFFFFFFFFFF;
+            return (bits < 0x7FF0000000000000) && (bits != 0) && ((bits & 0x7FF0000000000000) != 0);
```
Comment (jkotas) [inline src/mscorlib/src/System/Double.cs]: Is the `(bits != 0)` condition really needed here? 

`((bits & 0x7FF0000000000000) != 0)` implies that `bits != 0`. So it should be sufficient to write this as `(bits < 0x7FF0000000000000) && ((bits & 0x7FF0000000000000) != 0)`.

File: src/mscorlib/src/System/Single.cs [hunk 11]
```diff
@@ -40,38 +40,84 @@ public struct Single : IComparable, IFormattable, IConvertible
...
+        }
+
+        /// <summary>Determines whether the specified value is negative.</summary>
+        [Pure]
+        [System.Runtime.Versioning.NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe static bool IsNegative(float f)
         {
-            return *(int*)(&f) == 0x7F800000;
+            var bits = unchecked((uint)BitConverter.SingleToInt32Bits(f));
```
Comment (jkotas) [inline src/mscorlib/src/System/Single.cs]: This can be simplified to `return BitConverter.SingleToInt32Bits(f) < 0;`
